### ILAS: Introduction to Programming 2017/18
# Coursework Assignment: Plant-life Report

__Complete exercises A to D.__
<br>__The exercises should be completed using Python programming skills we have covered in class. The questions are focussed on an imaginary case study:__ 

>It is though that the acidification of an area of protected land is having a destructive effect on plant populations. 
<br>Experts are particularly worried about the demise of a species of shrub called *winter heath*, that supports the area's insect populations, and the spread of an acid-loving poisonous weed called *darley heath* . <br>Chemical waste from local industries are thought to be reposonsible for the soil acidification.
<br>Your job is to process data collected over a number of years to present as part of a report.
<br>The report will be used as evidence to try and impose restrictions disposal of industrial waste within the area.

<img src="img/map2.png" alt="Drawing" style="width: 500px;"/>

### Input data
Data collectd by a plant survey over the past 20 years is given in the folder `environmental_survey` in the `sample_data` folder of the ILAS_python repository.

The survey was conducted once a year. 

The locations and characteristics of plants and trees were recorded.

Soil pH was also recorded at different locations.

### Setting up

Create a new folder in which to store your project.

Copy the `environmental_survey` folder into the project folder.

### Part A: Assembling a Data Set

__Aim: Import plant data from .csv files and manipulate data to convert units and remove unecessary values.__

__(1.) Input and Output: Data Frames 
<br>*(5 marks)*__
<br>Write a Python program that imports the data from the file `plants2017` and stores it as a __`pandas DataFrame`__. 

The data set should contain only the data for shrub plants. 
<br>Remove the rows with "tree" in the plants column to leave only information about shrubs in your data set.

(Hint: After removing data from a DataFrame use `df.reset_index(drop=True)` (where 'df' is the DataFrame name) to re-assign index numbers).  

In [71]:
import pandas as pd
from pandas import read_csv
from pandas import DataFrame as df
from IPython.display import display

data=read_csv('sample_data/environmental_survey/plants2017.csv')#,index_col=0)
#print(data.keys())
#print(data)
data.drop(data.index[data.Plant =='tree'], inplace=True)
data.reset_index(drop=True)

GPS_lon    GPS_lat  Plant  height_m  leaf_length_cm  \
0     136.757630  35.732232  shrub      1.21             0.9   
1     136.758560  35.732232  shrub      1.34             6.1   
2     136.759490  35.732232  shrub      0.75             0.5   
3     136.760420  35.732232  shrub      1.26             0.5   
4     136.761350  35.732232  shrub      1.38             1.2   
5     136.762280  35.732232  shrub      1.07             3.4   
6     136.763210  35.732232  shrub      1.91             2.2   
7     136.764140  35.732232  shrub      2.06             0.5   
8     136.765070  35.732232  shrub      1.94             2.7   
9     136.766000  35.732232  shrub      1.06             2.4   
10    136.766930  35.732232  shrub      1.85             0.5   
11    136.768790  35.732232  shrub      0.80             8.9   
12    136.769720  35.732232  shrub      0.70             0.5   
13    136.771580  35.732232  shrub      1.13             4.1   
14    136.773439  35.732232  shrub      1.15             6.9   
15    136.774369  35.732232  shrub      1.72             2.4   
16    136.775299  35.732232  shrub      1.84             0.5   
17    136.776229  35.732232  shrub      0.76             2.2   
18    136.777159  35.732232  shrub      1.99             2.2   
19    136.778089  35.732232  shrub      1.28             0.5   
20    136.780879  35.732232  shrub      1.82             2.0   
21    136.781809  35.732232  shrub      0.69             0.5   
22    136.782739  35.732232  shrub      0.77             0.5   
23    136.784599  35.732232  shrub      0.57             6.3   
24    136.786459  35.732232  shrub      0.73             1.3   
25    136.787389  35.732232  shrub      1.78             0.5   
26    136.788318  35.732232  shrub      1.16            10.9   
27    136.790178  35.732232  shrub      1.13             4.2   
28    136.791108  35.732232  shrub      0.83             1.9   
29    136.792038  35.732232  shrub      0.75             9.3   
...          ...        ...    ...       ...             ...   
1451  136.821502  35.767771  shrub      0.80             1.8   
1452  136.822432  35.767771  shrub      0.85             0.5   
1453  136.824293  35.767771  shrub      0.68             5.1   
1454  136.825223  35.767771  shrub      0.69             3.7   
1455  136.826154  35.767771  shrub      1.87             0.9   
1456  136.827084  35.767771  shrub      1.71             4.2   
1457  136.828015  35.767771  shrub      0.68             1.4   
1458  136.828945  35.767771  shrub      1.87             5.2   
1459  136.830806  35.767771  shrub      1.14             3.9   
1460  136.834527  35.767771  shrub      1.26             3.8   
1461  136.836388  35.767771  shrub      0.86             2.6   
1462  136.837318  35.767771  shrub      0.68             4.9   
1463  136.838248  35.767771  shrub      1.77             0.6   
1464  136.839179  35.767771  shrub      0.63             5.1   
1465  136.840109  35.767771  shrub      0.74             0.8   
1466  136.841040  35.767771  shrub      0.74             5.3   
1467  136.841970  35.767771  shrub      1.14             0.5   
1468  136.843831  35.767771  shrub      1.01             6.1   
1469  136.844761  35.767771  shrub      0.66             3.1   
1470  136.845691  35.767771  shrub      0.54             5.7   
1471  136.846622  35.767771  shrub      1.70             0.5   
1472  136.848482  35.767771  shrub      0.57             5.9   
1473  136.849413  35.767771  shrub      0.69             1.0   
1474  136.850343  35.767771  shrub      0.88             0.5   
1475  136.851273  35.767771  shrub      1.87             1.7   
1476  136.852204  35.767771  shrub      0.69             0.5   
1477  136.854065  35.767771  shrub      0.50             3.5   
1478  136.854995  35.767771  shrub      0.63             6.2   
1479  136.855925  35.767771  shrub      0.50             7.0   
1480  136.856856  35.767771  shrub      0.71             4.1   

      leaf_aspect_ratio  bud_length_cm  

__(2.) Functions__
<br>__*(5 marks)*__
<br>The GPS location information for each plant is in units of decimal degrees. 
<br>To make them more "human readable", the values should be converted to represent each data point on a 2D grid, with units of metres (or kilometres). 

<img src="img/lat_long.png" alt="Drawing" style="width: 400px;"/>

The following equations can be used to approximate:
 - the vertical distance from the *equator* from `GPS_lat`
 - the horizontal distance from the *meridian* from `GPS_lon`
 
The latitude in m from the equator:

$lat = \frac{40,008,000 \times GPS_{lat}}{360} $


The longitude in m from the meridian:

$lon = \frac{40,075,160 \times GPS_{lon}}{360} \times \cos(GPS_{lat})$

<img src="img/ParametricCircle.png" alt="Drawing" style="width: 200px;"/>

Write code to convert GPS_lat and GPS_lon in decimal degrees to units of m or km, using the equation above. 
<br>__*Hint: `GPS_lat` and `GPS_lat` are given in degrees, `numpy.cos` automatically applies to angles given in radians.*__

Encapsulate your code in a function so that it can be applied to any data frame.
(Hint: your function should take the columns of data frame to be converted as its arguments).

Show your function works by applying it to your data frame.

(You can also want to *rename* your column heading as they are no longer GPS coordinates.)

In [72]:
import numpy as np
def conv(glong,glat):
    return ((40075160*glong/360)*np.cos(glat*np.pi/180),(40008000*glat/360))
data.GPS_lon,data.GPS_lat=conv(data.GPS_lon,data.GPS_lat)
display(data)

(12358033.527164182, 3971042.0496000005)


GPS_lon       GPS_lat  Plant  height_m  leaf_length_cm  \
0     1.235803e+07  3.971042e+06  shrub      1.21             0.9   
1     1.235812e+07  3.971042e+06  shrub      1.34             6.1   
2     1.235820e+07  3.971042e+06  shrub      0.75             0.5   
3     1.235829e+07  3.971042e+06  shrub      1.26             0.5   
4     1.235837e+07  3.971042e+06  shrub      1.38             1.2   
5     1.235845e+07  3.971042e+06  shrub      1.07             3.4   
6     1.235854e+07  3.971042e+06  shrub      1.91             2.2   
7     1.235862e+07  3.971042e+06  shrub      2.06             0.5   
8     1.235871e+07  3.971042e+06  shrub      1.94             2.7   
9     1.235879e+07  3.971042e+06  shrub      1.06             2.4   
10    1.235887e+07  3.971042e+06  shrub      1.85             0.5   
12    1.235904e+07  3.971042e+06  shrub      0.80             8.9   
13    1.235913e+07  3.971042e+06  shrub      0.70             0.5   
15    1.235929e+07  3.971042e+06  shrub      1.13             4.1   
17    1.235946e+07  3.971042e+06  shrub      1.15             6.9   
18    1.235955e+07  3.971042e+06  shrub      1.72             2.4   
19    1.235963e+07  3.971042e+06  shrub      1.84             0.5   
20    1.235971e+07  3.971042e+06  shrub      0.76             2.2   
21    1.235980e+07  3.971042e+06  shrub      1.99             2.2   
22    1.235988e+07  3.971042e+06  shrub      1.28             0.5   
25    1.236013e+07  3.971042e+06  shrub      1.82             2.0   
26    1.236022e+07  3.971042e+06  shrub      0.69             0.5   
27    1.236030e+07  3.971042e+06  shrub      0.77             0.5   
29    1.236047e+07  3.971042e+06  shrub      0.57             6.3   
31    1.236064e+07  3.971042e+06  shrub      0.73             1.3   
32    1.236072e+07  3.971042e+06  shrub      1.78             0.5   
33    1.236081e+07  3.971042e+06  shrub      1.16            10.9   
35    1.236097e+07  3.971042e+06  shrub      1.13             4.2   
36    1.236106e+07  3.971042e+06  shrub      0.83             1.9   
37    1.236114e+07  3.971042e+06  shrub      0.75             9.3   
...            ...           ...    ...       ...             ...   
1977  1.235829e+07  3.974992e+06  shrub      0.80             1.8   
1978  1.235837e+07  3.974992e+06  shrub      0.85             0.5   
1980  1.235854e+07  3.974992e+06  shrub      0.68             5.1   
1981  1.235862e+07  3.974992e+06  shrub      0.69             3.7   
1982  1.235871e+07  3.974992e+06  shrub      1.87             0.9   
1983  1.235879e+07  3.974992e+06  shrub      1.71             4.2   
1984  1.235887e+07  3.974992e+06  shrub      0.68             1.4   
1985  1.235896e+07  3.974992e+06  shrub      1.87             5.2   
1987  1.235913e+07  3.974992e+06  shrub      1.14             3.9   
1991  1.235946e+07  3.974992e+06  shrub      1.26             3.8   
1993  1.235963e+07  3.974992e+06  shrub      0.86             2.6   
1994  1.235971e+07  3.974992e+06  shrub      0.68             4.9   
1995  1.235980e+07  3.974992e+06  shrub      1.77             0.6   
1996  1.235988e+07  3.974992e+06  shrub      0.63             5.1   
1997  1.235997e+07  3.974992e+06  shrub      0.74             0.8   
1998  1.236005e+07  3.974992e+06  shrub      0.74             5.3   
1999  1.236013e+07  3.974992e+06  shrub      1.14             0.5   
2001  1.236030e+07  3.974992e+06  shrub      1.01             6.1   
2002  1.236039e+07  3.974992e+06  shrub      0.66             3.1   
2003  1.236047e+07  3.974992e+06  shrub      0.54             5.7   
2004  1.236055e+07  3.974992e+06  shrub      1.70             0.5   
2006  1.236072e+07  3.974992e+06  shrub      0.57             5.9   
2007  1.236081e+07  3.974992e+06  shrub      0.69             1.0   
2008  1.236089e+07  3.974992e+06  shrub      0.88             0.5   
2009  1.236097e+07  3.974992e+06  shrub      1.87             1.7   
2010  1.236106e+07  3.974992e+06  shrub      0.69             0.5   
2012  1.2

__(3.) Functions and Data Structures: Boolean Indexing__ 
<br>__*(5 marks)*__
<br>When fully grown, the four main shrubs that grow in the area can be identified by distinct features.

To include *only fully grown* plants in your data set:
- Write a function that selects only plants above a height of 50cm.
- Apply the function to your data set.
- Edit your function so that the same function may be used to:
  - remove plants below 50cm by default
  - remove plants below a height set by the user
  


### Part B: Refining the Data Set and Mapping pH

__Aim: Split the area over which the survey was taken into a grid of equally sized cells. Sort the pH samples by grid cell to show how pH varies across the area.__

__(1.) Input and Output__ 
<br>__*(2 marks)*__
<br>In the same Python file you wrote in __Part A__, import the data from the file `pH_2017` and store it as a new __`pandas DataFrame`__ called `pH`.
<br>

__(2.) Functions__
<br>__*(2 marks)*__
<br>Use the function that you wrote in __Part A (2.)__ to convert the the columns GPS_lat and GPS_lon in `pH` to units of m or km. 

SyntaxError: EOL while scanning string literal (<ipython-input-2-6d4107f8d662>, line 1)

The sampled area measures approximately 3445m x 3950m.
<br>An orthoganol grid of 15 x 15 cells (3000m x 3000m) can be used to represent the sampled area:
 - the grid is chosen to be slightly smaller than the sampled area so that no unsampled regions are included.
 - the origin is chosen to be at 
  - $x = x_{min} + \frac{3445-3000}{2}$
  - $y =  y_{min} + \frac{3950-3000}{2}$
  
<img src="img/map.png" alt="Drawing" style="width: 500px;"/>

The following equation can be used to map a point, $P$, in range A to range B.

$P_B=\frac{P_A-A_{min}}{A_{max}-A_{min}} \times (B_{max}-B_{min}) + B_{min}$

__(3.) Functions and mathematical operators.__ 
<br>__*(5 marks)*__

Write a function called `scale` to map points in the range (origin, origin+3000) to the range (0, 3000).

By floor dividing (seminar 2) points in the range 0 to 3000 by 200, each point can be assigned an integer value in the range 0 to 14. Create an additional step in your function that uses floor division to assign an x and y grid reference to each data point.  

Note:
- some grid references may be outside of the range 0 to 14.
- multiple cells will blong to the same grid reference.

Add two new columns to your DataFrame to store the x and y grid reference for each data point 

Store your function that assigns a grid index as function so that it can be applied to any data set collected in the same area.

__(3.) `numpy` multi-dimensional arrays.__ 
<br>__*(2 marks)*__
<br>_Find the mean of the pH readings taken in each grid cell.
<br>Use a 2D numpy array to store each mean reading at each 2D grid location. 



__(4.) Plotting__
<br>__*(3 marks)*__
<br>Plot the mean pH for each grid cell as a colour map of the gridded area. 
<br>You may use a *2D colour map* or a *3D plot*. 
<br>Save your figure as a .png file in your project folder.


### Part C: Classifying Plants Using Simple Mathematical Operations

__Aim: Sort the plant samples species. Produce a total count of each species in each grid cell.__

<br>The shrub plants in your DataFrame from __Part A__ can be catagorsied as one of four species.

The *average* physical characteristics of each *plant species* are shown in the table below:

|Shrub       |Height (m)|Leaf length (cm)|Leaf aspect ratio|Bud length (cm)|
|------------|----------|----------------|-----------------|---------------|
|Winter heath|       1.2|             3.5|              2.0|            2.3|              
|Bell heather|       1.8|             1.5|              1.2|            2.3|
|Brush bush  |       0.7|             2.1|             10.2|            1.5|
|Darley heath|       0.7|             2.2|              3.1|            1.7|

<br>The *vector quantisation algorithm* is a simple algorithm used for catagorisation.

It determines which catagory a data point should belong to from its closest proximity to a set of values representing possible catagories.
<br>Each value represents the *average* of the corresponding catagory.

The *closeness* of the characteristics of a point $(c_1, c_2, c_3, ... c_n)$ to the average value of a catagory $(ca_1, ca_2, ca_3, ... ca_n)$ can be determined by the magnitude:

<br>$d = \sqrt{(ca_1-c_1)^2 + (ca_2-c_2)^2 + (ca_3-c_3)^2 + ... + (ca_n-c_n)^2}$ <br>

If $d$ is evaluated for each catagory, the catagory with the *minimium* value of $d$ represents the closest fit.

The vector quantisation algorithm can be applied to each data point using a for loop or numpy broadcasting.



__(1.) Mathematical compuation with Numpy__ 
<br>__*(5 marks)*__
<br>Use the vector quantisation algorithm to determine the species of each plant.
<br>Hint: Use a for loop or use broadcasting. 
<br>Add a column to your DataFrame called "species" with the species of each plant that most closely fits the plant characteristics. 

__(2.) Functions__ 
<br>__*(1 mark)*__
<br>Use the function that you wrote for __Part B: (2.)__ to assign a grid reference to each data point.  <br>Save the grid refernce x and y value as two columns in your Data Frame.


__(3.) Data Structures: Lists__ 
<br>__*(5 marks)*__

Create a list for each of the following fields.
1. x grid index 
1. y grid index 
1. average pH reading
1. total count of *Winter heath* plant
1. total count of *Bell heather* plant
1. total count of *Brush bush* plant
1. total count of *Darley heath* plant

Loop through each grid cell and store a computed value for each field. 

Store the lists as a list of lists (nested lists).




### Part D: Using Multiple Files to Produce Time-Series Data

__Aim: Run all the steps that you coded in Part A-C for every envioronmental survey collected between the years 1997-2017 to produce time series data of the plant count and average pH.__

__(1.) Control Flow__ 
<br>__*(5 marks)*__
<br>Use a for loop to store a list of lists like you created in __Part C: (3.)__ for each year of the environmental survey (1997-2017)

Hint: You can loop through each plant survey using:
>```Python 
annual_data=[]
for year in range(1997, 2018):
    df = pd.read_csv("environmental_survey/plants" + str(year) + ".csv") 
```

Hint: Append the list of lists created in __Part C: (3.)__ to the list `annual_data` each time the code loops.


>```Python 
annual_data=[]
for year in range(1997, 2018):
    df = pd.read_csv("environmental_survey/plants" + str(year) + ".csv") 
```

__(2.) Plotting and Curve Fitting__
<br>__*(5 marks)*__

<br>The two closest industrial sites to the area of land are:
<br>__Sketchy inc.__ , established 1995, GPS coordinates lon = 136.7647, lat = 35.7336
<br>__Philamore co.__ , established 1990, GPS coordinates lon = 136.8262, lat = 35.7498

<br>Choose one grid cell that is close to an industrial site and one grid cell that is far from the industrial sites.
<br>Plot a scatter graph of the average pH and plant count for each species (y axis) against time (x axis).
<br>Fit a trendline to each data series
<br>Show the equation of the trendline and the proximity to an industrial site as labels. 